howso/feature_residuals.amlg

```diff
@@ -1102,6 +1191,152 @@
+	;  for calculations If an assoc, adds back in removed features as keys and copies the value of the primary key for each shared deviation group as its value.
+	;       e.g. {"a": 2.5, "b": 0} -> {"a": 2.5 "b":9 "c": 2.5 "d": 2.5} if features a, c, d were in a shared deviations group.
+	#!ExpandForSharedDeviations
+	(declare
+		(assoc compressed_values (null))
+		(if (size !sharedDeviationsMap)
+			(if (~ (list) compressed_values)
+				(assign (assoc
+					compressed_values (values (append compressed_values (indices !sharedDeviationsMap)) (true))
+				))
+				;if the compressed_values is not an list, it is an assoc.
+				(let
+					(assoc
+						;features that were dropped and had their values combined into their primary shared deviations feature
+						removed_features
+							(filter
+								(lambda
+									(not (contains_value
+										(indices compressed_values)
+										(current_value)
+									))
+								)
+								(indices !sharedDeviationsMap)
+							)
+					)
+					(assign (assoc
+						compressed_values
+							(append
+								compressed_values
+								;grab the values for the removed features by looking up their primary key values
+								(map
+									(lambda
+										(get compressed_values (get !sharedDeviationsMap (current_index)))
+									)
+									(zip removed_features)
+								)
+							)
+					))
+				)
+			)
+		)
+		compressed_values
+	)
+
+	;Helper function to prep shared feature deviations, by taking in a list of features and their list of residuals,
+	; combining all of the residuals under the primary feature key of the group of the shared deviations, and removing the features
+	; from the features list whose residuals have been combined into another features.
+	;parameters:
+	; features_list: A features list
+	; feature_residuals_lists: A features residuals whose lists are in the order of the features_list
+	#!PrepSharedDeviations
+	(declare
+		(assoc
+			features_list (assoc)
+			feature_residuals_lists (assoc)
+		)
+
+		(declare (assoc
+			joined_original_features_residuals_list (zip features_list feature_residuals_lists)
+		))
+
+		;assoc of features and their residuals that are not in a shared deviations group
+		(declare (assoc
+			non_shared_deviations_features_list
+				(remove joined_original_features_residuals_list (indices !sharedDeviationsMap))
+		))
+
+		;assoc of primary shared deviations features and their residuals that are combined from all the other features in their shared deviations group
+		(declare (assoc
+			shared_deviations_residuals_list_combined
+				(map
+					(lambda
+						(let
+							(assoc
+								;the group of features in the shared deviations group
+								feature_group (call !GetSharedDeviationGrouping (assoc feature_group_to_retrieve (current_index 2)))
+								current_feature (current_index 1)
+							)
+							(apply "append"
+								(map
```

Comment:
this entire map statement could be simplified into:
`(filter (unzip joined_original_features_residuals_list feature_group))`